SupyML/plugin.py

```
@@ -181,27 +189,55 @@ def _processLoop(self, node, variables):
+        loopContent = '<echo>%s</echo>' % loopContent
+        if loopType == 'onceif':
+            try:
+                if utils.str.toBool(self._parse(loopCond, variables
+                                                  ).split(': ')[-1]):
+                    output += self._parse(loopContent, variables) or ''
+            except AttributeError: # toBool() failed
+                pass
+            except ValueError: # toBool() failed
+                pass
         if loopType == 'while':
             try:
-                while utils.str.toBool(self._parse(loopCond, variables,
+                while utils.str.toBool(self._parse(loopCond, variables
                                                   ).split(': ')[-1]):
-                    loopContent = '<echo>%s</echo>' % loopContent
-                    output += self._parse(loopContent) or ''
+                    output += self._parse(loopContent, variables) or ''
             except AttributeError: # toBool() failed
                 pass
             except ValueError: # toBool() failed
                 pass
+        if loopType == 'foreach':
+            try:
+                tokens = callbacks.tokenize(self._parse(loopCond, variables).split(': ')[-1])
+                for token in tokens:
+                    variables.update({'loop': token})
+                    output += self._parse(loopContent, variables) or ''
+            except AttributeError:
+                pass
+            except ValueError:
+                pass
+        if loopType == 'range':
+            try:
+                left = int(self._parse(loopCond, variables).split(': ')[-1])
+                for token in range(left):
+                    variables.update({'loop': str(token)})
+                    output += self._parse(loopContent, variables) or ''
+            except AttributeError: # int() failed
+                pass
+            except ValueError: # int() failed
+                pass
```

please keep the scope of `try` blocks limited so they don't eat unexpected exceptions. For example:

```py
             loopCondValue = self._parse(loopCond, variables).split(': ')[-1]
             try:
                 left = int(loopCondValue)
            except (AttributeError, ValueError):
                pass
            else:
                for token in range(left):
                    variables.update({'loop': str(token)})
                    output += self._parse(loopContent, variables) or ''
```

---

done